FFmpeg discovery on Windows

FFmpeg provides [pre-built binaries for Windows](https://ffmpeg.org/download.html#build-windows). GIFT-Grab's FFmpeg discovery script however relies on pkg-config. For the experimental Windows branch, it'd be good to amend the discovery script in a very simple way to be able to link with the pre-built binaries.

FFmpeg might be needed to [speed up the RGB-to-BGRA conversion for BGRA captures with Epiphan](https://github.com/gift-surg/GIFT-Grab/issues/46#issuecomment-435311962)

Downloading the 4.1.3 pre-built dev binaries and running

`cmake -A x64 -D CMAKE_BUILD_TYPE=Release -D ZLIB_ROOT="%zlib_root%" -D USE_EPIPHAN_DVI2PCIE_DUO=ON -D ENABLE_NONFREE=ON -D USE_EPIPHANSDK=ON -D PYTHON_INCLUDE_DIRS="%python_include_dirs%" -D PYTHON_LIBRARY="%python_library%" -D BOOST_INCLUDEDIR="%boost_includedir%" -D BOOST_LIBRARYDIR="%boost_librarydir%" -D BUILD_PYTHON=ON -D USE_NUMPY=ON -D FFmpeg_LIBRARY_DIRS="C:/FFmpeg/binaries/ffmpeg-4.1.3-win64-dev/lib" -D FFmpeg_LIBS=avformat,avcodec,avutil,swscale,avfilter -D FFmpeg_INCLUDE_DIRS="C:/FFmpeg/binaries/ffmpeg-4.1.3-win64-dev/include" -D USE_FILES=ON -D USE_HEVC=ON -D ENABLE_NONFREE=ON -D USE_NVENC=ON ../../GIFT-Grab/src`

produces #87 related problems on Windows:

```
[...]
C:\Users\ds18\ws\GIFT-Grab\src\ffmpeg\ffmpeg_video_target.cpp(98): error C2065: 'not': undeclared identifier [C:\Users\ds18\ws\_b\gift-grab\giftgrab.vcxproj]
C:\Users\ds18\ws\GIFT-Grab\src\ffmpeg\ffmpeg_video_target.cpp(98): error C2146: syntax error: missing ')' before identifier '_codec' [C:\Users\ds18\ws\_b\gift-grab\giftgrab.vcxproj]
C:\Users\ds18\ws\GIFT-Grab\src\ffmpeg\ffmpeg_video_target.cpp(98): error C2059: syntax error: ')' [C:\Users\ds18\ws\_b\gift-grab\giftgrab.vcxproj]
[...]
```

--- a/src/cmake/ffmpeg/CMakeLists.txt
+++ b/src/cmake/ffmpeg/CMakeLists.txt
@@ -15,7 +15,8 @@ if(UNIX)
         CACHE INTERNAL "FFmpeg library dirs"
     )
 elseif(WIN32)
-    SET(FFmpeg_LIBS avformat avcodec avutil swscale avfilter
+    LIST(APPEND FFmpeg_LIBS avformat avcodec avutil swscale avfilter)
+    SET(FFmpeg_LIBS ${FFmpeg_LIBS}
         CACHE INTERNAL "FFmpeg libs"
     )
 endif()

--- a/src/cmake/ffmpeg/CMakeLists.txt
+++ b/src/cmake/ffmpeg/CMakeLists.txt
@@ -1,15 +1,19 @@
 CMAKE_MINIMUM_REQUIRED(VERSION 3.2.0)
 
-# Augment module path to be able to find needed packages
-LIST(APPEND CMAKE_MODULE_PATH ${CMAKE_CURRENT_SOURCE_DIR})
+if(UNIX)
+    # Augment module path to be able to find needed packages
+    LIST(APPEND CMAKE_MODULE_PATH ${CMAKE_CURRENT_SOURCE_DIR})
 
-FIND_PACKAGE(FFmpeg REQUIRED)
-SET(FFmpeg_INCLUDE_DIRS ${FFmpeg_INCLUDE_DIRS}
-    CACHE INTERNAL "FFmpeg include dirs"
-)
-SET(FFmpeg_LIBS ${FFmpeg_LIBS}
-    CACHE INTERNAL "FFmpeg libs"
-)
-SET(FFmpeg_LIBRARY_DIRS ${FFmpeg_LIBRARY_DIRS}
-    CACHE INTERNAL "FFmpeg library dirs"
-)
+    FIND_PACKAGE(FFmpeg REQUIRED)
+    SET(FFmpeg_INCLUDE_DIRS ${FFmpeg_INCLUDE_DIRS}
+        CACHE INTERNAL "FFmpeg include dirs"
+    )
+    SET(FFmpeg_LIBS ${FFmpeg_LIBS}
+        CACHE INTERNAL "FFmpeg libs"
+    )
+    SET(FFmpeg_LIBRARY_DIRS ${FFmpeg_LIBRARY_DIRS}
+        CACHE INTERNAL "FFmpeg library dirs"
+    )
+elseif(WIN32)
+    # TODO
+endif()

--- a/src/cmake/ffmpeg/CMakeLists.txt
+++ b/src/cmake/ffmpeg/CMakeLists.txt
@@ -15,5 +15,7 @@ if(UNIX)
         CACHE INTERNAL "FFmpeg library dirs"
     )
 elseif(WIN32)
-    # TODO
+    SET(FFmpeg_LIBS avformat avcodec avutil swscale avfilter
+        CACHE INTERNAL "FFmpeg libs"
+    )
 endif()

--- a/src/cmake/ffmpeg/CMakeLists.txt
+++ b/src/cmake/ffmpeg/CMakeLists.txt
@@ -15,8 +15,7 @@ if(UNIX)
         CACHE INTERNAL "FFmpeg library dirs"
     )
 elseif(WIN32)
-    LIST(APPEND FFmpeg_LIBS avformat avcodec avutil swscale avfilter)
-    SET(FFmpeg_LIBS ${FFmpeg_LIBS}
+    SET(FFmpeg_LIBS avformat avcodec avutil swscale avfilter
         CACHE INTERNAL "FFmpeg libs"
     )
 endif()

--- a/src/ffmpeg/ffmpeg_video_source.cpp
+++ b/src/ffmpeg/ffmpeg_video_source.cpp
@@ -75,7 +75,7 @@ bool VideoSourceFFmpeg::get_frame_dimensions(int & width, int & height)
 {
     std::lock_guard<std::mutex> buffer_lock_guard(_buffer_lock);
 
-    if (_avframe_processed->width > 0 and _avframe_processed->height > 0)
+    if (_avframe_processed->width > 0 && _avframe_processed->height > 0)
     {
         width = _avframe_processed->width;
         height = _avframe_processed->height;
@@ -115,7 +115,7 @@ bool VideoSourceFFmpeg::get_frame(VideoFrame & frame)
     }
     while (_avpacket.size > 0);
     av_packet_unref(&orig_pkt);
-    if (not success)
+    if (!success)
         return false;
 
     // push obtained frame down the pipeline
@@ -126,7 +126,7 @@ bool VideoSourceFFmpeg::get_frame(VideoFrame & frame)
 
     // pull processed frame from the pipeline
     ret = av_buffersink_get_frame(_pipeline_end, _avframe_processed);
-    if (ret == AVERROR(EAGAIN) or ret == AVERROR_EOF or ret < 0)
+    if (ret == AVERROR(EAGAIN) || ret == AVERROR_EOF || ret < 0)
         return false;
 
     // get non-aligned data from decoded frame
@@ -150,7 +150,7 @@ double VideoSourceFFmpeg::get_frame_rate()
 
     int num = _avformat_context->streams[_avstream_idx]->avg_frame_rate.num;
     int den = _avformat_context->streams[_avstream_idx]->avg_frame_rate.den;
-    if (den == 0 or num == 0)
+    if (den == 0 || num == 0)
         return 0.0;
     return static_cast<double>(num) / den;
 }
@@ -293,20 +293,20 @@ bool VideoSourceFFmpeg::ffmpeg_reset_pipeline(
     int x, int y, int width, int height
 )
 {
-    if (_pipeline_begin != nullptr and
+    if (_pipeline_begin != nullptr &&
         _pipeline_end != nullptr)
     // i.e. modifying existing pipeline
     {
-        if (x == _x and y == _y and
-            width == _avframe_processed->width and
+        if (x == _x && y == _y &&
+            width == _avframe_processed->width &&
             height == _avframe_processed->height)
             // no need to change anything
             return false;
     }
 
     // sanity checks in case need to (re)set pipeline
-    if (x < 0 or y < 0 or width < 0 or height < 0 or
-        x + width > _avformat_context->streams[_avstream_idx]->codec->width or
+    if (x < 0 || y < 0 || width < 0 || height < 0 ||
+        x + width > _avformat_context->streams[_avstream_idx]->codec->width ||
         y + height > _avformat_context->streams[_avstream_idx]->codec->height)
         throw VideoSourceError("Improper cropping parameters");
 
@@ -321,7 +321,7 @@ bool VideoSourceFFmpeg::ffmpeg_reset_pipeline(
     // allocate pipeline elements
     AVFilterInOut * out = avfilter_inout_alloc();
     AVFilterInOut * in  = avfilter_inout_alloc();
-    if (out == nullptr or in == nullptr)
+    if (out == nullptr || in == nullptr)
     {
         error_msg.append("Could not allocate FFmpeg pipeline IO points")
                  .append(get_ffmpeg_error_desc(AVERROR(ENOMEM)));
@@ -389,8 +389,8 @@ bool VideoSourceFFmpeg::ffmpeg_reset_pipeline(
     }
 
     // add cropping if necessary
-    if (x > 0 or y > 0 or
-        width < _avformat_context->streams[_avstream_idx]->codec->width or
+    if (x > 0 || y > 0 ||
+        width < _avformat_context->streams[_avstream_idx]->codec->width ||
         height < _avformat_context->streams[_avstream_idx]->codec->height)
     {
         ret = snprintf(pipeline_desc, sizeof(pipeline_desc),
@@ -471,13 +471,13 @@ bool VideoSourceFFmpeg::ffmpeg_realloc_proc_buffers(
     int ret = 0;
     std::string error_msg = "";
 
-    if (width <= 0 or height <= 0)
+    if (width <= 0 || height <= 0)
         throw VideoSourceError("Frame width and height"
                                " must be positive");
 
     if (_avframe_processed != nullptr)
     {
-        if (_avframe_processed->width == width and
+        if (_avframe_processed->width == width ||
             _avframe_processed->height == height)
             return false;
         av_frame_free(&_avframe_processed);
@@ -549,8 +549,8 @@ int VideoSourceFFmpeg::ffmpeg_decode_packet(AVPacket & avpacket,
         if (got_frame)
         {
             if (avframe->width != _avformat_context->streams[_avstream_idx]->codec->width
-                or avframe->height != _avformat_context->streams[_avstream_idx]->codec->height
-                or avframe->format != _avformat_context->streams[_avstream_idx]->codec->pix_fmt)
+                || avframe->height != _avformat_context->streams[_avstream_idx]->codec->height
+                || avframe->format != _avformat_context->streams[_avstream_idx]->codec->pix_fmt)
                 return -1;
             decoded = ret;
         }
@@ -558,7 +558,7 @@ int VideoSourceFFmpeg::ffmpeg_decode_packet(AVPacket & avpacket,
 
     /* If we use frame reference counting, we own the data and need
      * to de-reference it when we don't use it anymore */
-    if (got_frame and use_refcount)
+    if (got_frame && use_refcount)
         av_frame_unref(avframe);
 
     return decoded;

--- a/src/ffmpeg/ffmpeg_video_source.h
+++ b/src/ffmpeg/ffmpeg_video_source.h
@@ -1,3 +1,5 @@
+#pragma once
+
 #include "ivideosource.h"
 #include "broadcastdaemon.h"
 extern "C"

--- a/src/ffmpeg/ffmpeg_video_source.h
+++ b/src/ffmpeg/ffmpeg_video_source.h
@@ -1,5 +1,6 @@
 #pragma once
 
+#include "exports.h"
 #include "ivideosource.h"
 #include "broadcastdaemon.h"
 extern "C"
@@ -16,7 +17,7 @@ namespace gg
 //! \brief Uses FFmpeg for streaming video from
 //! compatible sources such as video files
 //!
-class VideoSourceFFmpeg : public IVideoSource
+class GG_EXPORTS VideoSourceFFmpeg : public IVideoSource
 {
 protected:
     //!

--- a/src/ffmpeg/ffmpeg_video_target.cpp
+++ b/src/ffmpeg/ffmpeg_video_target.cpp
@@ -81,7 +81,7 @@ void VideoTargetFFmpeg::init(const std::string filepath, const float framerate)
         throw VideoTargetError("Negative fps does not make sense");
     _framerate = framerate;
 
-    if (_codec_name == _CODEC_NAME_HEVC_X265 or
+    if (_codec_name == _CODEC_NAME_HEVC_X265 ||
         _codec_name == _CODEC_NAME_HEVC_NVENC)
         check_filetype_support(filepath, "mp4");
     else if (_codec_name == _CODEC_NAME_VP9_LIBVPX)
@@ -95,7 +95,7 @@ void VideoTargetFFmpeg::init(const std::string filepath, const float framerate)
         throw VideoTargetError("Could not allocate output media context");
 
     _codec = avcodec_find_encoder_by_name(_codec_name.c_str());
-    if (not _codec)
+    if (!_codec)
         throw VideoTargetError("Codec not found");
 
     _stream = avformat_new_stream(_format_context, _codec);
@@ -105,7 +105,7 @@ void VideoTargetFFmpeg::init(const std::string filepath, const float framerate)
 
     // allocate FFmpeg frame
     _frame = av_frame_alloc();
-    if (not _frame)
+    if (!_frame)
         throw VideoTargetError("Could not allocate video frame");
     _first_frame = true;
 }
@@ -162,7 +162,7 @@ void VideoTargetFFmpeg::ffmpeg_frame(const unsigned char * data,
     if (_framerate <= 0)
         throw VideoTargetError("Video target not initialised");
 
-    if (not frame)
+    if (!frame)
         throw VideoTargetError("FFmpeg frame not initialised");
 
     // return value buffers
@@ -436,7 +436,7 @@ void VideoTargetFFmpeg::encode_and_write(AVFrame * frame, int & got_output)
 void VideoTargetFFmpeg::finalise()
 {
     // Issue #130
-    if (not _is_finaliseable)
+    if (!_is_finaliseable)
         return;
 
     /* This condition means that append

--- a/src/ffmpeg/ffmpeg_video_target.cpp
+++ b/src/ffmpeg/ffmpeg_video_target.cpp
@@ -181,7 +181,12 @@ void VideoTargetFFmpeg::ffmpeg_frame(const unsigned char * data,
 //        _stream->codec->bit_rate = 400000;
         _stream->codec->width = width;
         _stream->codec->height = height;
-        _stream->time_base = (AVRational){ 1000, static_cast<int>(1000 * _framerate) };
+        /* the following causes "error C4576: a parenthesized type followed by an initializer
+         * list is a non-standard explicit type conversion syntax" on Windows:
+         * _stream->time_base = (AVRational){ 1000, static_cast<int>(1000 * _framerate) };
+         */
+        _stream->time_base.num = 1000;
+        _stream->time_base.den = static_cast<int>(1000 * _framerate);
         _stream->codec->time_base = _stream->time_base;
         _stream->codec->gop_size = 12;
         /* TODO emit one intra frame every twelve frames at most */

--- a/src/ffmpeg/ffmpeg_video_target.h
+++ b/src/ffmpeg/ffmpeg_video_target.h
@@ -1,5 +1,6 @@
 #pragma once
 
+#include "exports.h"
 #include "ivideotarget.h"
 
 extern "C" {
@@ -12,7 +13,7 @@ extern "C" {
 namespace gg
 {
 
-class VideoTargetFFmpeg : public IVideoTarget
+class GG_EXPORTS VideoTargetFFmpeg : public IVideoTarget
 {
 protected:
     //!

